File: C/E3_=_RSA-RAW/rsa_raw.c
// -*- coding: utf-8 -*-

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <gmp.h>

mpz_t clair, chiffre, dechiffre, e, n, d;           // Variables globales 

void fabrique(void){             // Fabrique d'une paire de clefs RSA (A MODIFIER)
  mpz_set_str(n, "196520034100071057065009920573", 10);
  mpz_set_str(e, "7", 10);
  mpz_set_str(d, "56148581171448620129544540223", 10);
}                                                

int main(void){
  mpz_inits(clair, chiffre, dechiffre, (void *)NULL);
  mpz_init(n);                  // Le module de la clef publique
  mpz_init(e);                  // L'exposant de la clef publique
  mpz_init(d);                  // L'exposant de la clef privée  

  mpz_set_str(clair, "4b594f544f", 16);

  /* Affichage du clair à l'aide de la fonction gmp_printf() */
  gmp_printf("Clair             : %Zd\n", clair);

  fabrique();

  /* Affichage des clefs utilisées à l'aide de la fonction gmp_printf() */
  gmp_printf("Clef publique (n) : %Zd\n", n);
  gmp_printf("Clef publique (e) : %Zd\n", e);
  gmp_printf("Clef privée (d)   : %Zd\n", d);

  /* On effectue d'abord le chiffrement RSA du clair avec la clef publique */
  mpz_powm(chiffre, clair, e, n);                    // Calcul du clair chiffré
  gmp_printf("Chiffré           : %Zd\n", chiffre);  

  /* On déchiffre ensuite avec la clef privée */
  mpz_powm(dechiffre, chiffre, d, n);              // Calcul du clair déchiffré
  gmp_printf("Clair déchiffré    : %Zd\n", dechiffre);

  mpz_clears(clair, chiffre, dechiffre, n, e, d, (void *)NULL) ;
  exit(EXIT_SUCCESS);
}  

/*
  $ make
  gcc -o clefs -I/usr/local/include -I/usr/include clefs.c -L/usr/local/lib -L/usr/lib -lgmp -std=c99 -Wall
  $ ./clefs 
  Code clair        : 323620918351
  Clef publique (n) : 196520034100071057065009920573
  Clef publique (e) : 7
  Clef privée (d)   : 56148581171448620129544540223
  Code chiffré      : 34528838076398914398154896650
  Code déchiffré    : 323620918351
 */

File: JAVA/E3_=_RSA-RAW/RSA_raw.java
// -*- coding: utf-8 -*-

import java.math.BigInteger;

public class RSA_raw
{
  private static BigInteger clair, chiffré, déchiffré ;
  private static BigInteger n ;      // Le module de la clef publique
  private static BigInteger e ;      // L'exposant de la clef publique
  private static BigInteger d ;      // L'exposant de la clef privée
  private static  KeyPair keyPair ;
    
  static void fabrique() {           // Fabrique d'une paire de clefs RSA (A MODIFIER)
    //n = new BigInteger("196520034100071057065009920573", 10);
    //e = new BigInteger("7", 10);
    //d = new BigInteger("56148581171448620129544540223", 10);
    BigInteger p=Alea.getPrimeNumber();
    BigInteger q=Alea.getPrimeNumber();
    n = p.multiply(q);
    BigInteger w = (p.subtract(BigInteger.ONE)).multiply(q.subtract(BigInteger.ONE));
    int result = 0;
    d=Alea.getPrimeNumber();
    while(result!= -1){
      d=  Alea.getPrimeNumber() ;
      result =d.compareTo(w) ;
    }
    e = d.modInverse(w);
    PublicKey publicKey = new PublicKey(n,e);
    PrivatKey privatKey = new PrivatKey(d);
    keyPair = new KeyPair(publicKey,privatKey);
    }

  public static void main(String[] args)
  {  
    clair = new BigInteger("4b594f544f", 16);

    /* Affichage du clair */
    System.out.println("Clair             : " + clair);
    
    fabrique(); 

    /* Affichage des clefs utilisées */
    System.out.println("Clef publique (n) : " + n);
    System.out.println("Clef publique (e) : " + e);
    System.out.println("Clef privée (d)   : " + d);

    /* On effectue d'abord le chiffrement RSA du clair clair avec la clef publique */
    chiffré = clair.modPow(e, n);
    System.out.println("Chiffré           : " + chiffré);

    /* On déchiffre ensuite avec la clef privée */
    déchiffré = chiffré.modPow(d, n);
    System.out.println("Déchiffré         : " + déchiffré);
  }
}

